### PR TITLE
feat: Breadcrumbs and events for child processes

### LINKED ISSUE #335: Submit events for `child-process-gone` events
Electron >= v11 has the [`child-process-gone`](https://www.electronjs.org/docs/api/app#event-child-process-gone) event. It would be useful to optionally generate sentry events for this.

### PR DESCRIPTION
**Pull request by timfish**
Closes #335

## Electron Normalisation
- Adds `onWebContentsCreated` and `onBrowserWindowCreated` to work around Electron v2 issue in a single place
- Add type-safe process exit reasons
- Improved `onRendererProcessGone` 
   - Allow hooking specific reasons
   - For older electron, fallback to previous limited events
   - Handle case where versions using breakpad incorrectly report `killed` for `crashed`
- Adds `onChildProcessGone` 
   - Allow hooking specific reasons
   - For older electron, fallback to previous GPU only events

## `SentryMinidump`
- Removes `addBreadcrumb` because this is now added in `ChildProcess`
- Check for and upload minidumps for crashed child processes
- Only changes the session to crashed if a minidump was found
- Add test with the GPU process minidump

## `ChildProcess`
- Add `ChildProcess` integration which is enabled by default
- Capture events for desired exit reasons (defaults to `['abnormal-exit', 'launch-failed', 'integrity-failure']`) 
- Add breadcrumbs for desired exit reasons (defaults to all)

Also fixes an issue with reporting missing count of test events

### CODE REVIEW

File: src/main/electron-normalize.ts [hunk 1]
```diff
@@ -32,27 +32,117 @@ export function supportsFullProtocol(): boolean {
...
+export const CRASH_REASONS: ExitReason[] = ['crashed', 'oom'];
+export const ALL_REASONS: ExitReason[] = [
+  'clean-exit',
+  'abnormal-exit',
+  'killed',
+  'crashed',
+  'oom',
+  'launch-failed',
+  'integrity-failure',
+];
```
Comment (AbhiPrasad) [inline src/main/electron-normalize.ts:53]:
```suggestion
export const EXIT_REASONS = [
  'clean-exit',
  'abnormal-exit',
  'killed',
  'crashed',
  'oom',
  'launch-failed',
  'integrity-failure',
] as const;

export type ExitReason = typeof EXIT_REASONS[number];

export const CRASH_REASONS = [EXIT_REASONS[3], EXIT_REASONS[4]] as const;
```

File: src/main/integrations/child-process.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      level = Severity.Warning;
+      break;
+    case 'crashed':
+    case 'oom':
+    case 'launch-failed':
+    case 'integrity-failure':
+      level = Severity.Critical;
+  }
+
+  return { message, level };
```
Comment (AbhiPrasad) [inline src/main/integrations/child-process.ts:38]: this is a nit, so feel free to ignore. IMO the immutable returns are clearer to understand.

```suggestion
  switch (reason) {
    case 'abnormal-exit':
    case 'killed':
      return { message, level: Severity.Warning};
    case 'crashed':
    case 'oom':
    case 'launch-failed':
    case 'integrity-failure':
      return { message, level: Severity.Critical};
    
    return return { message, level: Severity.Debug};
  }

  return { message, level };
```

File: src/main/integrations/child-process.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
+import { addBreadcrumb, captureMessage, getCurrentHub } from '@sentry/core';
+import { NodeClient } from '@sentry/node';
+import { Integration, Severity } from '@sentry/types';
+
+import { ALL_REASONS, ExitReason, onChildProcessGone, onRendererProcessGone } from '../electron-normalize';
+import { ElectronMainOptions } from '../sdk';
+
+interface ChildProcessOptions<T> {
```
Comment (AbhiPrasad) [inline src/main/integrations/child-process.ts:8]: Why make this generic? It's only be used for ExitReasons right?

File: src/main/integrations/child-process.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            category: 'child-process',
+            ...getMessageAndSeverity(details.reason, details.type),
+            data: details,
+          });
+        }
+      });
+
+      onRendererProcessGone(allReasons, (contents, details) => {
+        const { reason } = details;
+        const getName = (): string => options?.getRendererName?.(contents) || 'renderer';
```
Comment (AbhiPrasad) [inline src/main/integrations/child-process.ts:93]: Why is `getName` a function?

**Review by AbhiPrasad — APPROVED** [review]
cool that we can track the gpu stuff!

File: src/main/integrations/child-process.ts (hunk 2 above)
Comment (timfish) [reply src/main/integrations/child-process.ts:38]: No I agree here!

File: src/main/integrations/child-process.ts (hunk 4 above)
Comment (timfish) [reply src/main/integrations/child-process.ts:93]: Yeah the name will be required at this point, so no reason to lazily evaluate it

File: src/main/electron-normalize.ts (hunk 1 above)
Comment (timfish) [reply src/main/electron-normalize.ts:53]: Not sure I agree with this one. 

The array indexing makes it easier to break in the future and it's not immediately clear which events are crashes. 

They both transpile down to a similar length JavaScript.

File: src/main/integrations/child-process.ts (hunk 3 above)
Comment (timfish) [reply src/main/integrations/child-process.ts:8]: watch me try and attempt that across PRs 🤣

File: src/main/electron-normalize.ts (hunk 1 above)
Comment (AbhiPrasad) [reply src/main/electron-normalize.ts:53]: yeah I was iffy about the clarity from `CRASH_REASONS` as well. I think the `type ExitReason` is good though.

we can just make it

```ts
export const CRASH_REASONS: Readonly<ExitReason[]> = ['crashed', 'oom'] as const;
```

File: src/main/integrations/child-process.ts (hunk 3 above)
Comment (AbhiPrasad) [reply src/main/integrations/child-process.ts:8]: lol

We can just do it in another PR, np. Step by step!

File: src/types.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;
+
+type OrBool<T> = {
```
Comment (AbhiPrasad) [inline src/types.ts:3]: These don't have to be exported?

**Review by AbhiPrasad — APPROVED** [review]

File: src/types.ts (hunk 5 above)
Comment (timfish) [reply src/types.ts:3]: Apparently not!

Comment (AbhiPrasad) [reply src/types.ts:3]: 😮

**Pull request merged**